package.json

```diff
@@ -104,6 +92,43 @@
   "main": "./dist/date-picker/app-date-picker.js",
   "module": "./dist/date-picker/app-date-picker.js",
   "typings": "./dist/date-picker/app-date-picker.d.ts",
+  "typesVersions": {
```

Comment:
@userquin Could you please explain why is this required?

IIUC, `exports` is more than enough for TypeScript to read the corresponding types correctly from this [article](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-9.html#exports-is-prioritized-over-typesversions).

Reply:
If your tsconfig file has `compilerOptions.moduleResolution="node"` TS will be using node10, there shoud be dts files in root folder and this package doesn't include them, types version entry will allow TS to detect dts files.

Reply:
> TS will be using node10, there shoud be dts files in root folder and this package doesn't include them, types version entry will allow TS to detect dts files.

I don't remember that's how TypeScript works. IIRC, it reads from `types` or `typings` in the `package.json` and you can see that I already it defined.

Reply:
That's when importing app-datepicker, the problem when importing any app-datepicker subpackage.

Reply:
@userquin I looked into this and I need your help to clarify some things.

```ts
import { AppDatePicker } from 'app-datepicker/app-date-picker';
```

This import statement uses the `exports` field in `package.json`, which is only supported by `node16` and later versions. To make this work in `node10`, we can use the `typesVersions` field to ship different types for different TypeScript versions.

IMHO, this does not feel right to me to use `typesVersions` for such use case because `node10` users should avoid this kind of import and instead use relative paths to point to a specific file, like this:

```ts
// instead of doing this
// import { AppDatePicker } from 'app-datepicker/app-date-picker';

// node10 users should do this
import { AppDatePicker } from 'app-datepicker/dist/date-picker/app-date-picker.js';
```

WDYT?

Reply:
You need support for node, node16(esm) and bundler.
The problem is in d.ts files, we will need to generate d.mts files for node16: when using relative imports inside a d.ts you need the extension and so you cannot reuse the same d.ts for node, node16 and bundler.
I'm going to try to use tsup and unbuild to generate proper d.ts and d.mts files.
In node16 you need to use `.js` extensions inside the d.ts (m.dts).

Reply:
Check this huge PR migrating to ESM-Only UnoCSS packages: 
https://github.com/unocss/unocss/pull/3380

Reply:
I kind of get it but this project does not support CommonJS/ require at all. It aims to ship only ESM files because it is supposed to run only on the browsers at the end of the day. Generating types just to support node10 and CommonJS is definitely out of the scope of the project unless it is a typing bug/ issue. TypeScript using `moduleResolution=node` can still reference the correct file and typing file without any issue as long as they use the relative file path. `moduleResolution=node16|bundler` and ESM are what this project supports and anything other than that is no-no. Hope it makes sense.

Reply:
I didn't mention CJS at all ;). If you only support node16 and bundler you only need to fix typings.ts modules in this PR, and we can also remove types versions (node 10) from package.json file.